Add REST API method for suggestions for a batch of documents

juhoinkinen opened the pull request:
Adds a new `/v1/projects/{project_id}/suggest-batch` REST API method. Based on the branch of PR #663, implements parts of #579.

The method accepts the documents with optional ids and and optional paramaters json:
```json
{
  "documents": [
    {
      "id": "doc-1234",
      "text": "A quick brown fox jumped over the lazy dog."
    }
  ],
  "parameters": {
    "language": "string",
    "limit": 10,
    "threshold": 0
  }
}
``` 
Response is (id is null if the document in request does not have one):
```json
[
  {
    "results": [
      {
        "label": "Archaeology",
        "notation": "42.42",
        "score": 0.85,
        "uri": "http://example.org/subject1"
      }
    ],
    "id": "doc-1234"
  }
]
```
The `id` field is similar to the `external_id` of [MonkeyLearn classifier](https://monkeylearn.com/api/v3/#classify) and to the `index` [AWS Comprehend BatchDetectKeyPhrases](https://docs.aws.amazon.com/comprehend/latest/APIReference/API_BatchDetectKeyPhrases.html).

osma commented:
Would it make sense to limit the number of documents in a single REST call to the minibatch size (currently 32)? After all, there would be little benefit (except maybe avoiding some HTTP overhead) in processing more than one minibatch on the backend side.

juhoinkinen commented:
I tried to reduce duplication in `annif.yaml` as much as possible. Unfortunately I did not find a way to define parameters for suggestions in only place, because for `/suggest` they are used in request body, whereas for `/suggest-batch` in query parameters.

juhoinkinen commented:
I reverted two refactoring commits. I think https://github.com/NatLibFi/Annif/commit/067fec38223adb1c1e957fe480ad9e6878e5147c should not have been problematic, but I run [openapi-diff tool](https://hub.docker.com/r/openapitools/openapi-diff/), and it complained about `POST   /projects/{project_id}/suggest`:
```
==========================================================================
==                            API CHANGE LOG                            ==
==========================================================================
                              Annif REST API                              
--------------------------------------------------------------------------
--                              What's New                              --
--------------------------------------------------------------------------
- POST   /projects/{project_id}/suggest-batch

--------------------------------------------------------------------------
--                            What's Changed                            --
--------------------------------------------------------------------------
- POST   /projects/{project_id}/learn
  Request:
        - Changed application/json
          Schema: Backward compatible
- POST   /projects/{project_id}/suggest
  Request:
        - Changed application/x-www-form-urlencoded
          Schema: Broken compatibility
          Changed property type:  (object -> object)
--------------------------------------------------------------------------
--                                Result                                --
--------------------------------------------------------------------------
                 API changes broke backward compatibility                 
--------------------------------------------------------------------------
```
Now after the reverts the tool still sees changed application/json in ` /projects/{project_id}/learn`, but I think it is wrong, at least I cannot see a difference in there, and the tool anyway says that the change is backward compatible.

juhoinkinen commented:
Finally I managed to drop the unnecessary merge and revert commits. I recreated the PR branch from the current main and then cherry-picked the good commits from a backup PR branch. I don't know why rebasing did not work: there ended up also all commits made to main not matter what I tried. Changing base branch back and forth did not help as usually.

But now there is a problem installing just released version of a dependency in GH Actions (but not on my laptop):

    Unable to find installation candidates for libclang (16.0.0)

juhoinkinen commented:
> But now there is a problem installing just released version of a dependency in GH Actions (but not on my laptop):
> 
> ```
> Unable to find installation candidates for libclang (16.0.0)
> ```

This seemed to be caused by issue https://github.com/sighingnow/libclang/issues/46, which got fixed.

osma commented on annif/openapi/annif.yaml:
```diff
@@ -96,23 +92,68 @@ paths:
+    post:
+      tags:
+      - Automatic subject indexing
+      summary: suggest subjects for given documents using batched backend operations
```
I think this is going into too much implementation detail. How about

    suggest subjects for the given batch of up to 32 documents

osma reviewed (changes requested):
I tested this locally and it seems to work well. A couple of points to consider:

1. I suggested a little change to the wording of the method summary
2. I verified that the method fails if given more than 32 documents, as it should. But the error message is a bit confusing:

```json
{
  "detail": "[{'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'plaa'}] is too long - 'documents'",
  "status": 400,
  "title": "Bad Request",
  "type": "about:blank"
}
```

Basically the "detail" field includes the whole request, which could be extremely long; and it ends with "is too long - 'documents'" which is not that helpful. Is this something we could change or is this coming directly from Connexion so we can't do anything about it? I would like to see a more helpful message, which wouldn't include the whole request body, just a message stating that there were too many documents.

3. Related to the above, can we write (for example using schemathesis) unit tests that verify that the API accepts 32 documents, but doesn't accept 33 documents? I'm worried that if we change the implementation later, the limit of 32 will be dropped and it could lead to problems on the backend side.

If you can fix at least some of the above it would be great, but if points 2 and/or 3 are too difficult, I think it's also OK to merge this as it is.

juhoinkinen commented:
>  1. I suggested a little change to the wording of the method summary
 

Done.

>    2. I verified that the method fails if given more than 32 documents, as it should. But the error message is a bit confusing:
> 
> 
> ```json
> {
>   "detail": "[{'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'Olipa kerran'}, {'text': 'plaa'}] is too long - 'documents'",
>   "status": 400,
>   "title": "Bad Request",
>   "type": "about:blank"
> }
> ```
> 
> Basically the "detail" field includes the whole request, which could be extremely long; and it ends with "is too long - 'documents'" which is not that helpful. Is this something we could change or is this coming directly from Connexion so we can't do anything about it? I would like to see a more helpful message, which wouldn't include the whole request body, just a message stating that there were too many documents.
> 

I added `CustomRequestBodyValidator` in `annif/openapi/validation.py` module. It is a child class of the connexion `RequestBodyValidator`, and it overrides the [default `validate_schema()` method](https://github.com/spec-first/connexion/blob/2.14.2/connexion/decorators/validation.py#L207) to modify the message in the "detail" field to only `validation error: too many items - 'documents'`.

This seems to be the [recommended way to modify the validation](https://connexion.readthedocs.io/en/stable/request.html#custom-validators), found it via https://github.com/spec-first/connexion/issues/558.

>   3. Related to the above, can we write (for example using schemathesis) unit tests that verify that the API accepts 32 documents, but doesn't accept 33 documents? I'm worried that if we change the implementation later, the limit of 32 will be dropped and it could lead to problems on the backend side.
> 

I added tests for cases 32 and 33 documents, and restored the manually written Swagger/OpenAPI tests that I removed in #682. These test do not rely on Schemathesis but on the `app_client` fixture, and this allows to check which error code is given for which error.

juhoinkinen commented on annif/openapi/validation.py:
```diff
@@ -0,0 +1,44 @@
+        """Validate the request body against the schema."""
+
+        if self.is_null_value_valid and is_null(data):
+            return None
```
I just marked the line with `# noqa`. The validation mechanism in [soon-to-be-released Connexion 3](https://github.com/spec-first/connexion/releases) is changed (https://github.com/spec-first/connexion/pull/1610), so this needs to be addressed again when/if upgrading to Connexion 3.
